api docs: Document the /users/me/<stream_id>/topics endpoint.

YagoGG opened the pull request:
**Testing Plan:** Made sure the docs display correctly and `test-api` passed.

YagoGG commented:
Renaming this as WIP since I don't know what's happening with the tests yet.

I also have the feeling that the changes in `requirements/` have to be improved somehow. Maybe merging all the deps that are duplicate in `common`, `dev` and `prod` into `common`? Only modifying `common.in` wasn't causing any effects when I tried it, so I guess that `dev.txt` overrides the deps we have there.

eeshangarg commented on templates/zerver/api/get-stream-topics.md:
```diff
@@ -0,0 +1,70 @@
+Get all the topics in a specific stream
+
+`GET {{ api_url }}/v1/users/me/<stream_id>/topics`
+
```
There needs to be an **Arguments** section that documents all the potential arguments that can be passed to this API endpoint. Use the `generate_api_arguments_table` Markdown extension and `templates/zerver/api/arguments.json` to document the arguments. :)

YagoGG replied:
I just didn't add that because this endpoint in particular doesn't require any arguments (besides the `<stream_id>` which is included in the URL). Should I add it anyways?

eeshangarg replied:
@YagoGG: Yeah I think we should add it. It is a good thing we can be consistent across all of our API docs and document all arguments as exhaustively as possible. :)

eeshangarg commented on templates/zerver/api/get-stream-topics.md:
```diff
@@ -0,0 +1,70 @@
+
+{generate_code_example|get-stream-topics|fixture}
+
+An example of a JSON response for when the user is not authorized
```
`An example JSON response...` is better/more concise than `An example of a JSON response...`

eeshangarg commented:
@timabbott, @YagoGG: It may not be an internal mypy thing, but that's just what happened to break when I tested this locally. I can't see why this PR would cause the build failures here though.

YagoGG commented:
Updated the PR after addressing your comments, @eeshangarg. Thanks for reviewing!

eeshangarg commented on templates/zerver/api/arguments.json:
```diff
@@ -286,5 +286,13 @@
+    "get-stream-topics.md": [
+        {
+            "argument": "stream_id",
+            "description": "The ID of the stream whose topics should be fetched.",
```
I think just `The unique ID of the stream` should be fine. :)

eeshangarg commented on templates/zerver/api/get-stream-topics.md:
```diff
@@ -0,0 +1,76 @@
+
+## Arguments
+
+**Note:** these arguments go in the endpoint's URL, not as a querystring.
```
This line is misleading, "go in the endpoint's URL" and "as a querystring" are the same thing, right? Either way, I think it should be `**Note**: The following arguments are all URL query parameters.`. :)

YagoGG replied:
It isn't the same. An argument can be either in the URL's path (like in this case):

    chat.zulip.org/api/v1/users/me/THE_STREAM_ID_GOES_HERE/topics

Or as a querystring:

    chat.zulip.org/api/v1/some/endpoint?some_argument=some_value

These aren't interchangeable, and in this endpoint you should use the former.

If you think that a change in the phrasing can make this clearer I'll be happy to push an update :)

eeshangarg replied:
@YagoGG: Oops my bad, sorry! Okay, well then let's just leave that `Note` part out. I feel like the examples should be enough for the reader to draw a distinction between how to pass in the argument (based on whether one is using Python, JS or curl).

Also, you were right that the endpoint doesn't technically take in any arguments but the Python code that interfaces with the endpoint *does* take in an argument and that's why I thought we should document it explicitly. Just thought I'd clarify! :)

YagoGG commented:
Updated the PR to fix the requirements as discussed in Zulip.

timabbott commented:
The mypy error is interesting; I'm debugging.

timabbott commented:
I fixed the mypy thing (was unrelated consequence of the upgrade).  I also merged the first two commits.  You should do a separate commit for the upgrade, since that's mergable now (and it'd be nice to separate it from the rest).

I can't merge the new `/topics` docs because the `max_message_id` values are randomly determined every time one does `populate_db`, so that test won't pass for other folks' systems.  I think we'll need to do a mock or something in the API test for that feature to work around the issue...

timabbott commented:
I think probably we just need to have the test be tolerant of any int in those positions, rather than requiring the exact value.  That, or the test needs to do a database query to get the right answers.

@eeshangarg any ideas on how to configure that behavior?

YagoGG commented:
I was thinking about something similar to the DB query, but using the bindings themselves: using them to fetch the corresponding IDs, and then modify the fixture after loading it into memory.

Anyways, is there any reason in special why we don't make `do-destroy-rebuild*` always produce the same result?

timabbott commented:
Basically, the content of the development environment test messages is randomly generated because you kinda want some variability.  We maintain some consistency on a given machine, but it's painful to manage keeping the randomization constant as we expand what goes in the dev environment.

I think we could potentially have a deterministic template database used for this test (I think we basicall do that with that backend and casper tests...).

timabbott commented on zerver/tests/test_openapi.py:
```diff
@@ -0,0 +1,81 @@
+TEST_RESPONSE = '400'
+
+
+class OpenAPITest(ZulipTestCase):
```
What is this test class testing?  It feels like it's testing our OpenAPI parsing? 

Might be worth a docstring, and renaming it to `OpenAPIParsingTest`.

YagoGG replied:
I decided to rename it to `OpenAPIToolsTest`, since `zerver/lib/openapi.py` isn't exactly for parsing our OpenAPI content, but manage the stuff that has already been parsed (i.e. tools that help us handle OpenAPI stuff).

Let me know if you still prefer `OpenAPIParsingTest`.

timabbott replied:
that's fine.

timabbott commented on templates/zerver/api/get-all-streams.md:
```diff
@@ -79,7 +79,7 @@ A typical successful JSON response may look like:
-An example of a JSON response for when the user is not authorized
-to use the `include_all_active` parameter:
+An example JSON response for when the user is not authorized to use the
+`include_all_active` parameter:
```
This would benefit from e.g. "because they are not an organization administrator" or whatever the reason is.

timabbott commented on requirements/common.in:
```diff
@@ -169,7 +169,7 @@ pyoembed==0.1.2
 # release tarballs (which is a bug).  So we need to pin it makes sense
 # to pin a version from Git rather than a release.
--e "git+https://github.com/zulip/python-zulip-api.git@0.4.6#egg=zulip==0.4.6_git&subdirectory=zulip"
+-e "git+https://github.com/zulip/python-zulip-api.git@292a3bc0671dd28cde79dd532a86a3a049bfb2f5#egg=zulip&subdirectory=zulip"
```
Can you split the API version upgrade to its own commit?

I think @eeshangarg was having some trouble upgrading this due to some weird mypy error, so I'd like to be able to debug and fix that without reading the rest of this.

timabbott commented:
> We can bear with some duplication (like the code currently in this PR, that shows description both in the OpenAPI and Markdown files), or implement a reader for the responses like we've done with other fields (like the arguments, for instance).

My opinion is that our priority is to have decent docs, and using the OpenAPI format is very much secondary (in part because I'm not sure we'll ever end up displaying the OpenAPI files directly).  

But yes, if we wanted to share the descriptions, we'd want to do something like the fixture thing.

YagoGG commented:
Thanks for reviewing, @timabbott! I just updated the PR addressing your comments.

mypy does seem to be complaining about the `zulip` imports, both in code I've touched and code I haven't. I tried dropping the `python-zulip-api` upgrade commit just to check if it's what you were talking about, but it seems to be unrelated.

timabbott commented:
I merged the first two commits.  The last bit is blocked on that mypy thing, which I'm hoping to debug via  #9487.
